libraries/AC_PID/AC_PID_2D.cpp

```diff
@@ -133,16 +133,17 @@ void AC_PID_2D::update_i(const Vector2f &limit)
     _pid_info_y.limit = false;
 
     Vector2f delta_integrator = (_error * _ki) * _dt;
+    _integrator += delta_integrator;
     if (!is_zero(limit.length_squared())) {
-        // zero delta_vel if it will increase the velocity error
+        // do not let integrator increase in length if delta_integrator is in the direction of limit
         if (is_positive(delta_integrator * limit)) {
-            delta_integrator.zero();
+            float integrator_length = _integrator.length();
```

Comment:
It would definitely be good to test this change to AC_PID_2D with various values to ensure it is doing what is expected.  So for example, an AC_PID object could be temporarily created in the Copter 1hz loop and then various values passed in for the limit and then check that the resulting I-term is as expected.

These two lines seem a bit odd actually.. it seems that we take the length of the integrator and then immediately limit it to that length so this shouldn't have any effect right?  I think perhaps the idea was that we should take the length before the delta_integrator has been added?

Reply:
I guess there is no chance that we might suddenly shorten the integrator?  I think the intention is that we don't allow the integrator to grow in the direction of the limit.. but that's different than suddenly shortening it.

Reply:
Yes, this was a requested change to avoided calculating the length when we don't have to. We have to.